### ISSUE TITLE
Discussion: stanza order

### ISSUE THREAD

**Issue by jawshooah** [caskroom/homebrew-cask#16161]
Moving this discussion from #16158 to avoid polluting the PR with unrelated comments.

**Comment by adityadalal924** [comment]
A proposal:

```ruby
# version info block
:version
:shasum

# download link block
:url
:appcast, :shasum
:auto_updates

# basic info block
:name
:homepage
:license

# addon/helper blocks blocks
depends_on (as needed)
conflicts_with (as needed)
stage_only  (as needed)
accessibility_access  (as needed)

pre/post flights (as needed)

container

# artifact(s) block
:app, :pkg

# uninstall blocks
:uninstall

:zap
```

Not sure where `caveats` and `gpg` would best fit

**Comment by vitorgalvao** [comment]
For context: the current order of the first two “blocks” has (to a point) some established logic. It’s about what contributors should be concerned with. `version` and `shasum` are the *only* stanzas guaranteed to change with a version update, hence why they are separate and at the top. Next are `url` and `appcast` which can change but many times won’t (interpolation)<sup>1</sup>. Next are all the others that (in theory) won’t change over the lifetime of the app: `name`, `homepage`, `license`. `gpg` is (currently) placed under `url` for convenience (those stanzas generally use `#{url}`).

Then come the other blocks, that change widely across casks. In the past, we tried to juggle a bit semantic logic with updatability logic (the latter being more important), and that’s a stance I still agree with. To me, *the less common it is for an app to have a type of stanza, the further down it should be*. No reason way we can’t embed some operations order logic in there as well, though.

To touch briefly on `stage_only`, that is to be used when there isn’t an artifact stanza to use, so it would make sense to put it where those would be,

My proposal:

```
version
sha256

url
appcast,
  :sha256
name
homepage
license
auto_updates
accessibility_access
gpg, :key_id # on same line, since first part is typically small

suite
app
pkg
installer
binary
colorpicker
font
input_method
internet_plugin
prefpane
qlplugin
screen_saver
service
audio_unit_plugin
vst_plugin
artifact
stage_only

uninstall

zap

preflight

postflight

uninstall_preflight

uninstall_postflight

depends_on
conflicts_with

caveats
```

Some final considerations: every stanza that has options (symbols after a `,`) shall have them on separate lines, one per line. Exceptions are `gpg` (explained in proposal) and `:target`, when not applied to `url`<sup>2</sup>, for the same reason as `gpg` (very small lines, even with the option).

---

<sup>1</sup> Yes, `:sha256` in `appcast` will change, but those need to be coupled together, and `appcast` itself won’t change much.

<sup>2</sup> Not yet an option, but planned.

**Comment by Amorymeltzer** [comment]
:+1: to nearly all of @vitorgalvao's proposal.

>To me, the less common it is for an app to have a type of stanza, the further down it should be.

Love this.  I'd add in a corollary that casks should *very* roughly follow the order of operations, where doing so doesn't conflict immensely with the above.[1]  For these (and perhaps sentimental reasons), I do think `uninstall` and `zap` make sense appearing at the end.

I'd also be fine inserting a newline after the `url/appcast` in the "download link block" per @adityadalal924, though.  "Naming" those blocks is a great idea, that would be a wonderful thing to incorporate into [CONTRIBUTING.md](https://github.com/caskroom/homebrew-cask/blob/master/CONTRIBUTING.md)

---
1: A third law would be to favor required stanzas earlier than optional, but that will essentially boil down to what @vitorgalvao said.  Nice to have [three](https://en.wikipedia.org/wiki/Three_Laws_of_Robotics), though.

**Comment by vitorgalvao** [comment]
> I do think `uninstall` and `zap` make sense appearing at the end.

Agreed (with the exception of `caveats`, since it is purely informational and the most dispensable stanza, in a way). Moving `depends_on`/`conflicts_with` further up next to container also makes sense regarding order of operations, and bundling one-line stanzas together looks better.

As for `url` and `appcast`, at least to me those two lines “abandoned” so near the top look weird to me, especially if there is no `appcast`, where we’d have a weird case of:

```
version
sha256

url

name
homepage
```

And that doesn’t look good at all, to me.

The reason I wasn’t more strict with the order of operations regarding installation/uninstallation is at some point it actually looks a bit confusing/ugly:

```
preflight

app…

postflight

uninstall_preflight

uninstall

uninstall_postflight

zap
```

New proposal (I didn’t simply update the previous one since @Amorymeltzer’s agreement might no longer apply, making posting weird and hard to follow, and we’d lose important iterations on the process):

```
version
sha256

url
appcast,
  :sha256 # shown here since it is (planned to be) mandatory
name
homepage
license
auto_updates
accessibility_access
gpg, :key_id # on same line, since first part is typically small

container
depends_on
conflicts_with

suite
app
pkg
installer
binary
colorpicker
font
input_method
internet_plugin
prefpane
qlplugin
screen_saver
service
audio_unit_plugin
vst_plugin
artifact, :target # :target is (should be made) mandatory in this case
stage_only

preflight

postflight

uninstall_preflight

uninstall_postflight

uninstall

zap

caveats
```

**Comment by Amorymeltzer** [comment]
:100:

Repeating question added after-the-fact: Why `suite` before `app`?

**Comment by vitorgalvao** [comment]
> Why `suite` before `app`?

When you need to move a `suite`, you’re moving a directory of stuff, the core of the app and everything else, so it somehow made sense to me to have the bulk of it all at the top, since it seems to carry more weight than “all the little parts”. I could also see a logic for having it at the end of that list, but that would mean there would be so many unimportant stuff before it didn’t really make sense.

I don’t think we have a case with both, though, so it shouldn’t make any real difference.

**Comment by adityadalal924** [comment]
The bottom half looks good to me. For the top half, I propose:
```
version
sha256

url
appcast,
  :sha256 # shown here since it is (planned to be) mandatory
name
homepage
license
gpg, :key_id # on same line, since first part is typically small

auto_updates
accessibility_access
conflicts_with
depends_on


container
suite
{snip)
```

`auto_updates` and `accessibility_access` are "additional" things, and should be put with the other "extras". I also reordered them to alphabetical order for this section. I have no strong opinion on `gpg` and am fine with that anywhere.

I moved container to the next group as they are strategies to handle suites/apps/artifacts and should be grouped as such.

----------------

@Amorymeltzer Relevant XKCD: http://xkcd.com/1613/

**Comment by vitorgalvao** [comment]
> `auto_updates` and `accessibility_access` are "additional" things, and should be put with the other "extras".

But they are also “things that will never change”, which is why they were in the previous group. That said, I see the point in your logic. However, I disagree with container. `container` is also an additional thing, and one that is mostly a hack. More importantly, though, `container` isn’t mandatory for cask audit but at least one of the artifact stanzas is (will be). Artifacts are also the crux of casks, they’re *what you want*. Everything else is just cruft that exists to help you get the artifact in question, and putting `container` there is giving it way more importance than what it deserves. It is more akin to `depends_on`/`conflicts_with`, something that must be performed *before* we can get to the artifact.

Updated:

```
version
sha256

url
appcast,
  :sha256 # shown here since it is (planned to be) mandatory
name
homepage
license
gpg, :key_id # on same line, since first part is typically small

auto_updates
accessibility_access
conflicts_with
depends_on
container

suite
app
pkg
installer
binary
colorpicker
font
input_method
internet_plugin
prefpane
qlplugin
screen_saver
service
audio_unit_plugin
vst_plugin
artifact, :target # :target is (should be made) mandatory in this case
stage_only

preflight

postflight

uninstall_preflight

uninstall_postflight

uninstall

zap

caveats
```

**Comment by vitorgalvao** [comment]
However, that means that many times we’ll have an `auto_updates` stanza alone, which I’m not sure I like. `auto_updates` is and important stanza, though, since it dictates how `upgrade` will affect the cask, so I guess it’s fine for it to have a prominent position some times.

**Comment by adityadalal924** [comment]
Sounds good - I'm fine with your proposal above

**Comment by vitorgalvao** [comment]
@Amorymeltzer @jawshooah Are you in agreement as well?

**Comment by Amorymeltzer** [comment]
Yup
